File: src/leetcode/binary_search/374_guess_number_higher_or_lower.rb
# frozen_string_literal: true

# LeetCode 75
# https://leetcode.com/problems/guess-number-higher-or-lower/description/?envType=study-plan-v2&envId=leetcode-75

# Time complexity: O(log n)
# Space complexity: O(1)

# The guess API is already defined for you.
# @param num, your guess
# @return -1 if num is higher than the picked number
#          1 if num is lower than the picked number
#          otherwise return 0
# def guess(num)

def guess(num)
  picked = 6

  num <=> picked
end

def guess_number(range)
  left = 0
  right = range

  while left <= right
    pick = left + ((right - left) / 2)

    guess_check = guess(pick)

    return pick if guess_check.zero?

    if guess_check == 1
      left = pick + 1
    else
      right = pick - 1
    end
  end
end

File: src/leetcode/binary_search/374_guess_number_higher_or_lower.ts
// LeetCode 75
// https://leetcode.com/problems/guess-number-higher-or-lower/description/?envType=study-plan-v2&envId=leetcode-75

// Time complexity: O(log n)
// Space complexity: O(1)

/**
 * Forward declaration of guess API.
 * @param {number} num   your guess
 * @return 	  -1 if num is higher than the picked number
 *			      1 if num is lower than the picked number
 *            otherwise return 0
 * var guess = function(num) {}
 */

const guess = (num: number): number => {
  const picked = 6;

  if (num < picked) {
    return -1;
  } else if (num > picked) {
    return 1;
  } else {
    return 0;
  }
};

const guessNumber = (range: number): number => {
  let left = 0;
  let right = range;

  let pick = 0;

  while (left <= right) {
    pick = left + Math.floor((right - left) / 2);

    const guessCheck = guess(pick);

    if (guessCheck === 0) break;

    if (guessCheck === 1) {
      left = pick + 1;
    } else {
      right = pick - 1;
    }
  }

  return pick;
};

export default guessNumber;
